{% extends 'analysispage.html' %}

{% load static from staticfiles %}


{% block title %}Family Coverage Report &#187; <span>{{ gene.symbol }}</span>{% endblock %}

{% block links %}
    Family: <a href="{% url 'family_home' project_id=project.project_id family_id=family.family_id %}">{{ family }}</a>
{% endblock %}

{% block desc %}
    <p>
        This page contains details about the coverage across coding regions of this gene.
        For the purpose of coverage analysis, we categorize sequencing data as one of the following: [TODO: need more details]
        <ul>
            <li><strong>Callable:</strong> there was enough data to generate a variant call</li>
            <li><strong>Low Coverage:</strong> there were not enough reads in this region to generate a variant call</li>
            <li><strong>Poor Mapping:</strong> it was difficult to align reads to this region</li>
        </ul>
    </p>
    <p>
        The data below is organized by <em>coding region</em>.
        Here, the definition of a coding region is a region of the genome that has an exon that contributes to a protein coding exon.
        Since exons often overlap, asking for "the exons in a gene" can be misleading and/or confusing.
        So, we merge all overlapping exons into a single coding regions.
        However, a downside is that you can't refer to coding regions by number, as you would refer to "exon 8" of a gene.
        To map between an exon number and a coding region, you must cross reference their genomic coordinates.
    </p>
{% endblock %}

{% block innercontent %}

    {% include 'family_warnings.html' %}

    <div class="container">
        <div class="row">
            <div class="col-md-3">
                <dl class="well" >
                    <h4 class="section-header">Gene Info</h4>
                    <dt>Ensembl ID</dt>
                    <dd>{{ gene.gene_id }}</dd>
                    <dt>Chromosome</dt>
                    <dd>{{ gene.chr }}</dd>
                    <dt>Position</dt>
                    <dd>{{ gene.start }} - {{ gene.stop }}</dd>
                    <dt>Total Coding Size</dt>
                    <dd>{{ whole_gene.gene_coding_size }} bases</dd>
                    <p>More info can be found on the <a target="_blank" href="/quickgene/{{ gene.gene_id }}">gene page</a></p>
                </dl>
            </div>
            <div class="col-md-4">
                <h4 class="section-header">Total Coverage</h4>
                <p>Cumulative coverage across all regions and all individuals in this family. </p>
                <canvas id="all-exons-canvas" width="280" height="180"></canvas>
            </div>
            <div class="col-md-5">
                <h4 class="section-header">Summary By Individual</h4>
                <div id="coverage-by-individual-container"></div>
            </div>
        </div>
        <div class="row">
            <div class="col-md-12">
                <h4 class="section-header">Coverage By Coding Region</h4>
                <div id="coverage-by-exon-container"></div>
            </div>
        </div>

    <script type="text/template" id="tpl-single-exon-coverage">
    <div class="single-exon-coverage<% if (intervals) { %> gopopover<% } %>"
        <% if (intervals) { %>
            title="<%= indiv_id %>"
            data-content="<% if (coverage_list != undefined) { %>
                <% _.each(coverage_list, function(c) { %>
                    <%= c.xstart % 1e9 %> - <%= c.xstop % 1e9 %> <strong><%= c.xstop-c.xstart+1 %> <%= c.coverage %></strong><br/>
                <% }); %>
            <% } %>
            "
            data-html="true"
            data-placement="left"
        <% } %>
        >
        <% if (totals.ratio_callable == 1) { %>
        <i class="fa fa-check"></i>
        <% } else { %>
        <span class="ratio"><%= Math.round(totals.ratio_callable*100) %>%</span>
        <span class="callable"><%= totals.callable %></span> /
        <span class="low_coverage"><%= totals.low_coverage %></span> /
        <span class="poor_mapping"><%= totals.poor_mapping %></span>
        <% } %>
    </div>
    </script>

    <script type="text/template" id="tpl-coverage-by-individual">
    <table class="table coverage-by-exon-table">
        <thead>
            <tr>
                <th>Individual</th>
                <th>Coverage</th>
                <th>Completely Covered Regions</th>
            </tr>
        </thead>
        <tbody>
            <% _.each(indiv_ids, function(indiv_id) { %>
            <tr>
                <td><%= indiv_id %></td>
                <td><%= coverage_cell_template({totals: coverages[indiv_id].gene_totals, intervals:false}) %></td>
                <td>
                    <%= coverages[indiv_id].gene_totals.num_complete_regions %> /
                    <%= coverages[indiv_id].gene_totals.num_complete_regions+coverages[indiv_id].gene_totals.num_incomplete_regions %>
                </td>
            </tr>
            <% }); %>
        </tbody>
    </table>
    </script>

    <script type="text/template" id="tpl-coverage-by-transcript">
    <table class="table coverage-by-exon-table">
        <thead>
            <tr>
                <th>Transcript</th>
                <th>Total</th>
                <th>Biotype</th>
                <th>CDS Size</th>
                <th># Fully Covered</th>
            </tr>
        </thead>
        <tbody>
            <% _.each(by_transcript, function(transcript) { %>
            <tr>
                <td><%= transcript.transcript_id %></td>
                <td><%= coverage_cell_template({exon: transcript.totals, intervals:false}) %></td>
                <td><%= transcript.biotype %></td>
                <td><%= transcript.cds_size %></td>
                <td><%= transcript.totals.num_good %>/<%= transcript.num_cds %></td>
            </tr>
            <% }); %>
        </tbody>
    </table>
    </script>

    <script type="text/template" id="tpl-coverage-by-exon">
    <table class="table coverage-by-exon-table">
        <thead>
            <tr>
                <th>Start</th>
                <th>End</th>
                <th>Size</th>
                <% _.each(indiv_ids, function(indiv_id) { %>
                    <th><%= indiv_id %></th>
                <% }); %>
            </tr>
        </thead>
        <tbody>
            <% _.each(coding_regions, function(coding_region, i) { %>
            <tr>
                <td><%= coding_region.start %></td>
                <td><%= coding_region.stop %></td>
                <td><%= coding_region.size %></td>
                <% _.each(indiv_ids, function(indiv_id) { %>
                    <td>
                        <%= coverage_cell_template({
                            totals: coverages[indiv_id].coverage_specs[i].totals,
                            intervals: true,
                            coverage_list: coverages[indiv_id].coverage_specs[i].coverage_list,
                            indiv_id: indiv_id,
                        }) %>
                    </td>
                <% }); %>
            </tr>
            <% }); %>
        </tbody>
    </table>
    </script>

    {% include "javascript.html" %}

    <script type="text/javascript">
        var COVERAGES = {{ coverages_json|safe }};
        var WHOLE_GENE = {{ whole_gene_json|safe }};
        var CODING_REGIONS = {{ coding_regions_json|safe }};
        var INDIV_IDS = {{ indiv_ids_json|safe }};
    </script>
    <script type="text/javascript" src="{% static 'js/libs/rgraph/RGraph.common.core.js' %}"></script>
    <script type="text/javascript" src="{% static 'js/libs/rgraph/RGraph.common.dynamic.js' %}"></script>
    <script type="text/javascript" src="{% static 'js/libs/rgraph/RGraph.common.key.js' %}"></script>
    <script type="text/javascript" src="{% static 'js/libs/rgraph/RGraph.pie.js' %}"></script>
    <script type='text/javascript' src='{% static "js/coverage.js" %}'></script>

{% endblock %}

