{% load static from staticfiles %}

<script type="text/template" id="tpl-basic-variant">
    <div class="basicvariant">
        <div class="highlight-msg" style="font-size:12pt">
            <% console.log(variant); %>
            <% if(variant.extras && variant.extras.clinvar_clinsig) { %>
                <div style="display: inline-block;">
                    <% if(variant.extras.clinvar_variant_id) { %>
                        <a target="_blank" href="http://www.ncbi.nlm.nih.gov/clinvar/variation/<%= variant.extras.clinvar_variant_id %>">
                    <% } else if(variant.extras.clinvar_allele_id) { %>
                        <a target="_blank" href="http://www.ncbi.nlm.nih.gov/clinvar/?term=<%= variant.extras.clinvar_allele_id %>[alleleid]">
                    <% } else { %>
                        <a target="_blank" href="http://www.ncbi.nlm.nih.gov/clinvar">
                    <% } %>
                        ClinVar:
                        <span style="font-weight:500">
                            <% _.each(variant.extras.clinvar_clinsig.split(";"),
                                function(clinsig) {
                                    var color = utils.getClinvarClinsigColor(clinsig);
                            %>
                                <i style="color:<%= color %>"><%= clinsig %></i>
                            <% }); %>

                            <% if(variant.extras.clinvar_gold_stars !== undefined) { %>
                                &nbsp; &nbsp;
                                <% for(var i=0; i < 4; i+= 1) { %>
                                    <% if(i < variant.extras.clinvar_gold_stars) { %>
                                        <i class="icon-star" style="color: #FFB70A"></i>
                                    <% } else { %>
                                        <i class="icon-star" style="color: #BBBBBB"></i>
                                    <% } %>
                                <% } %>
                            <% } %>
                        </span>
                    </a>
                </div>
            <% } %>
            <% if(variant.extras && variant.extras.hgmd_class) {
                if(variant.extras.clinvar_clinsig) { %>
                   &nbsp; &nbsp;
                <% }
                var hgmd_color = utils.getHGMDClassColor(variant.extras.hgmd_class);
                var hgmd_label = utils.getHGMDClassName(variant.extras.hgmd_class); %>
                <div style="display: inline-block;">
                    <a target="_blank" href="https://portal.biobase-international.com/hgmd/pro/mut.php?acc=<%= variant.extras.hgmd_accession %>">
                        HGMD: <span style="font-weight:500">
                            <i style="color:<%= hgmd_color %>"><%= hgmd_label %></i>
                        </span>
                    </a>
                </div>
            <% } %>
            <% if (variant.extras && variant.extras.family_tags && variant.extras.family_tags.length > 0) { %>
                <div class="tags">
                    <div class="greytext" style="vertical-align:top; margin-right:50px"><b>Tags: </b></div><span style="display:inline-block">
                    <% _.each(variant.extras.family_tags, function(tag) { %>
                        <% if(show_tag_details) { %>
                            <span class="label" style="background-color:<%= tag.color %>; margin-left:10px;"><%= tag.tag %></span>
                            <i>
                                tagged by
                                <% if(tag.user) { %>
                                    <%= tag.user.display_name %>
                                <% } else { %>
                                    unknown user
                                <% } %>
                                <% if(tag.date_saved) { %> (<%= tag.date_saved %>) <% } %>
                            </i>
                            <% if(tag.search_url) { %>
                                <a style="margin-left:10px" href="<%= tag.search_url %>">
                                    <i class="fa fa-search" aria-hidden="true"></i>
                                </a>
                                <a href="<%= tag.search_url %>">re-run variant search</a>
			            <% } %>
			            <br />
                        <% } else { %>
                            <span class="label" style="background-color:<%= tag.color %>;"><%= tag.tag %></span>
                        <% } %>
		            <% }); %>
		            </span>
                </div>
            <% } %>
            <% if (allow_functional && variant.extras.family_functional_data.length > 0) { %>
                <div class="tags functional-data">
                    <div class="greytext" style="vertical-align:top; margin-right:50px"><b>Fxnl: </b></div><span style="display:inline-block">
                    <% _.each(variant.extras.family_functional_data, function(tag) { %>
                        <% if(show_tag_details) { %>
                            <span class="label" style="background-color:<%= tag.tag_config.color %>; margin-left:10px;"><%= tag.tag %></span>
                            <% if(tag.metadata) { %>
                                <b>&nbsp<%= tag.tag_config.metadata_title %>: <%= tag.metadata %>&nbsp</b>
                            <% } %>
                            <i>
                                tagged by
                                <% if(tag.user) { %>
                                    <%= tag.user.display_name %>
                                <% } else { %>
                                    unknown user
                                <% } %>
                                <% if(tag.date_saved) { %> (<%= tag.date_saved %>) <% } %>
                            </i>
                            <% if(tag.search_url) { %>
                                <a style="margin-left:10px" href="<%= tag.search_url %>">
                                    <i class="fa fa-search" aria-hidden="true"></i>
                                </a>
                                <a href="<%= tag.search_url %>">re-run variant search</a>
			            <% } %>
			            <br />
                        <% } else { %>
                            <span class="label" style="background-color:<%= tag.tag_config.color %>;"><%= tag.tag %></span>
                        <% } %>
		            <% }); %>
		            </span>
                </div>
            <% } %>
            <%  if (variant.extras && variant.extras.family_notes && variant.extras.family_notes.length > 0) { %>
                <div class="notes">
                    <div class="greytext"><b>Notes: </b></div>
                    <span style="display:inline-block">
                        <% for(var i = variant.extras.family_notes.length - 1; i >= 0; i--) {
                            var family_note = variant.extras.family_notes[i];
                            %>
                            <%= family_note.note %>
                            <i>by
                                <% if(family_note.user) { %>
                                    <%= family_note.user.display_name %>
                                <% } else { %>
                                    unknown user
                                <% } %>
                                <% if(family_note.submit_to_clinvar) { %>
                                    <span style="color:red"> for clinvar </span>
                                <% } %>
                                <% if(family_note.date_saved) { %>
                                    (<%= family_note.date_saved %>)
                                <% } %>
                            </i>
                            <a class="edit-variant-note" data-target="<%= family_note.note_id %>"><i class="fa fa-pencil" aria-hidden="true"></i></a>
                            <a class="delete-variant-note" data-target="<%= family_note.note_id %>"><i class="fa fa-trash-o"  aria-hidden="true"></i></a>
                            <br />
                        <% } %>
                    </span>
                </div>

            <% } %>
        </div>

        <% if (leftview) { %>
            <div class="leftview"></div>
        <% } %>

        <div class="cell icons" style="display:none;">
            <% if (variant.extras.disease_genes && variant.extras.disease_genes.length > 0 ) { %>
                <i class="fa fa-warning icon-popover"
                    title="Gene List"
                    data-content="<% _.each(variant.extras.disease_genes, function (a) { %><%= a %><% }); %>"></i>
            <% } %>
            <% if (variant.extras.in_disease_gene_db) { %>
                <i class="fa fa-plus icon-popover"
                    title="Present in Disease Database"
                    data-content="This variant is in a gene that has been linked to a disease phenotype.
                    Click the gene for more info. "></i>
            <% } %>
            <% if (variant.extras.family_notes && variant.extras.family_notes.length > 0 ) { %>
                <i class="fa fa-bookmark search-flag-icon"
                   data-xpos="<%= variant.xpos %>"
                   data-ref="<%= variant.ref %>"
                   data-alt="<%= variant.alt %>"></i>
            <% } %>
        </div>
        <% if(variant.annotation && (variant.annotation.main_transcript || variant.annotation.vep_annotation)) {
            var worst_vep_annotation = variant.annotation.main_transcript || variant.annotation.vep_annotation[variant.annotation.worst_vep_annotation_index];
        } else {
            var worst_vep_annotation = null;
        } %>
        <% if (show_gene) {  %>
            <div class="cell genes">
                <% var worst_gene_id = worst_vep_annotation && (worst_vep_annotation.gene || worst_vep_annotation.gene_id); %>
                <% var gene_id = worst_gene_id %>
                <% var gene = gene_id && variant.extras.genes[gene_id]; %>
                <% if(gene) { %>
                    <div class="gene-cell">
                        <a class="gene-link" data-gene_id="<%= gene_id %>"><%= gene.symbol || variant.extras.gene_names[gene_id] %></a><br/>
                        <sub>
                            <a href="http://www.gtexportal.org/home/gene/<%= gene.symbol %>" target="_blank">GTEx</a><br />
                            <a href="http://gnomad-beta.broadinstitute.org/gene/<%= gene.symbol %>" target="_blank">gnomAD</a><br />
                            <% if(show_gene_search_link && project_id) { %>
                                <a href="/project/<%= project_id %>/gene/<%= gene_id %>" target="_blank">Gene Search</a><br/>
                            <% } %>
                        </sub>
                        <div class="highlights">
                            <% if ((gene.missense_constraint && gene.missense_constraint_rank[0] < 1000) || (gene.tags && gene.tags.missense_constraint && gene.tags.missense_constraint_rank[0] < 1000)) { %>
                                <span class="label label-default" style='display:inline'>MISSENSE CONSTR
                                    <i class="icon-question-sign icon-popover"
                                       title="Missense Constraint"
                                       data-placement="right"
                                       data-content="This gene ranks <%= gene.missense_constraint_rank ? gene.missense_constraint_rank[0] : gene.tags.missense_constraint_rank[0] %> most constrained out of <%= gene.missense_constraint_rank ? gene.missense_constraint_rank[1] : gene.tags.missense_constraint_rank[1] %> genes under study in terms of missense constraint (z-score: <%= (gene.missense_constraint || gene.tags.missense_constraint).toPrecision(4) %>). Missense contraint is a measure of the degree to which the number of missense variants found in this gene in ExAC v0.3 is higher or lower than expected according to the statistical model described in [K. Samocha 2014]. In general this metric is most useful for genes that act via a dominant mechanism, and where a large proportion of the protein is heavily functionally constrained.">
                                    </i>
                                </span>
                                <br />
                            <% } %>
                            <% if ((gene.lof_constraint && gene.lof_constraint_rank[0] < 1000) || (gene.tags && gene.tags.lof_constraint && gene.tags.lof_constraint_rank[0] < 1000)) { %>
                                <span class="label label-default">
                                    LOF CONSTR
                                    <i class="icon-question-sign icon-popover"
                                       title="Loss of Function Constraint"
                                       data-placement="right"
                                       data-content="This gene ranks as <%= gene.lof_constraint_rank ? gene.lof_constraint_rank[0] : gene.tags.lof_constraint_rank[0] %> most intolerant of LoF mutations out of <%= gene.lof_constraint_rank ? gene.lof_constraint_rank[1] : gene.tags.lof_constraint_rank[1] %> genes under study. This metric is based on the amount of expected variation observed in the ExAC data and is a measure of how likely the gene is to be intolerant of loss-of-function mutations."></i>
                                </span><br/>
                            <% } %>
                        </div>
                        <% if (gene.disease_db_phenotypes && gene.disease_db_phenotypes.length > 0) { %>
                            <span class="label label-default">IN OMIM</span><br/>
                        <% } %>
                        <% if (gene.disease_gene_lists && gene.disease_gene_lists.length > 0 ) { %>
                            <% _.each(gene.disease_gene_lists, function (gene_list_name) { %>
                                <span class="label label-danger icon-popover"
                                      title="Gene List"
                                      data-content="<%= gene_list_name %>">
                                    GENE LIST: <%= gene_list_name.substring(0,6) %> <%= gene_list_name.length > 6 ? '..' : '' %>
                                </span><br/>
                            <% }); %>
                        <% } %>
                    </div>
                <% } %>
                <% if(Object.keys(variant.extras.genes).length > 1) { %>
                    <hr style="margin: 10px 0px;" />
                <% } %>
                <% _.each(variant.extras.genes, function(gene, gene_id) { %>
                    <% if(gene_id != worst_gene_id) { %>
                        <div class="gene-cell icon-popover" style="font-family: inherit;" data-html="true" title="Gene Details"
                             data-content="
                                <div class='highlights'>
                                    <% if ((gene.missense_constraint && gene.missense_constraint_rank[0] < 1000) || (gene.tags && gene.tags.missense_constraint && gene.tags.missense_constraint_rank[0] < 1000)) { %>
                                        <span class='label label-default' style='display:inline'>MISSENSE CONSTRAINT</span>
                                        This gene ranks <%= gene.missense_constraint_rank ? gene.missense_constraint_rank[0] : gene.tags.missense_constraint_rank[0] %> most constrained out of <%= gene.missense_constraint_rank ? gene.missense_constraint_rank[1] : gene.tags.missense_constraint_rank[1] %> genes under study in terms of missense constraint (z-score: <%= (gene.missense_constraint || gene.tags.missense_constraint).toPrecision(4) %>). Missense contraint is a measure of the degree to which the number of missense variants found in this gene in ExAC v0.3 is higher or lower than expected according to the statistical model described in [K. Samocha 2014]. In general this metric is most useful for genes that act via a dominant mechanism, and where a large proportion of the protein is heavily functionally constrained.
                                        <br />
                                    <% } %>
                                    <% if ((gene.lof_constraint && gene.lof_constraint_rank[0] < 1000) || (gene.tags && gene.tags.lof_constraint && gene.tags.lof_constraint_rank[0] < 1000)) { %>
                                        <span class='label label-default'>LOF CONSTRAINT</span>
                                        This gene ranks as <%= gene.lof_constraint_rank ? gene.lof_constraint_rank[0] : gene.tags.lof_constraint_rank[0] %> most intolerant of LoF mutations out of <%= gene.lof_constraint_rank ? gene.lof_constraint_rank[1] : gene.tags.lof_constraint_rank[1] %> genes under study. This metric is based on the amount of expected variation observed in the ExAC data and is a measure of how likely the gene is to be intolerant of loss-of-function mutations.
                                        <br/>
                                    <% } %>
                                </div>
                                <% if (gene.disease_db_phenotypes && gene.disease_db_phenotypes.length > 0) { %>
                                    <span class='label label-default'>IN OMIM</span><br/>
                                <% } %>
                                <% if (gene.disease_gene_lists && gene.disease_gene_lists.length > 0 ) { %>
                                    <span class='label label-danger'>GENE LISTS:</span>
                                    <%= gene.disease_gene_lists.join(', ') %>
                                <% } %>
                            "
                        >
                            <a class="gene-link" style="font-size: 16px;color: grey;" data-gene_id="<%= gene_id %>">
                                <%= gene.symbol || variant.extras.gene_names[gene_id] %>
                            </a><br/>
                             <% if(variant.annotation && variant.annotation.vep_annotation) { %>
                                <% var major_consequence = null; %>
                                <% var major_consequence_rank = 1000; %>
                                <% _.each(variant.annotation.vep_annotation, function(annotation) { %>
                                    <% if(annotation.gene_id === gene_id && annotation.major_consequence_rank < major_consequence_rank) { %>
                                        <% major_consequence_rank = annotation.major_consequence_rank; %>
                                        <% major_consequence = annotation.major_consequence.replace(/_/g, ' '); %>
                                    <% } %>
                                <% }) %>
                                <small><%= major_consequence %></small><br/>
                            <% } %>
                        </div>
                    <% } %>
                <% }) %>
            </div>
        <% } %>
        <div class="cell location">
            <a class="external-link" href="<%= utils.getVariantUCSCBrowserLink(variant, variant.extras.genome_version) %>" target="_blank">
                <%= 'chr'+variant.chr %>:<%= variant.pos %>
            </a><br/>
            <span class="allele-display" title="<%= variant.ref %>"><%= variant.ref %></span>
            <i class="fa fa-angle-right"></i>
            <span class="allele-display" title="<%= variant.alt %>"><%= variant.alt %></span><br/>
            <% if (variant.annotation && variant.annotation.rsid) { %>
                <a class="external-link" target="_blank" href="http://www.ncbi.nlm.nih.gov/SNP/snp_ref.cgi?searchType=adhoc_search&type=rs&rs=<%= variant.annotation.rsid %>">
                    <%= variant.annotation.rsid %>
                </a>
	        <% } %><br />
	        <% if (variant.extras && variant.extras.genome_version == "38" && variant.extras.grch37_coords) { %><br />
                <a class="external-link" href="<%= utils.getVariantUCSCBrowserLink(variant, '37') %>" target="_blank" >
                    <small>hg19: <%= variant.extras.grch37_coords.split("-").slice(0, 2).join(":") %></small>
                </a><br/>
	        <% } else if (variant.extras.genome_version == "38") { %>
                <small>hg19: liftover failed</small>
            <% } %>
            <div style="margin-top:10px;">
                <% if(family_read_data_is_available) { %>
                    <sup><a class="view-reads"><img src="{% static 'images/igv_reads_12x12.png' %}"/> &nbsp; SHOW READS</a></sup>
                <% } %>
            </div>
        </div>
        <div class="cell annotations">
            <a class="annotation-link"
                data-xpos="<%= variant.xpos %>"
                data-ref="<%= variant.ref %>"
                data-alt="<%= variant.alt %>">
                <% if (variant.annotation) { %>
                       <%= variant.annotation.vep_group.replace(/_/g, ' ') %>
                <% } %>
            </a>
            <% if(worst_vep_annotation) { %>
                <% if (worst_vep_annotation.lof == 'LC' || worst_vep_annotation.lof_flags == 'NAGNAG_SITE') {
                    var loftee_tooltip = '';
                    if (worst_vep_annotation.lof_filter != undefined && worst_vep_annotation.lof_filter != '') {
                        var lof_filters = _.uniq(worst_vep_annotation.lof_filter.split("&")).map(
                            function(lof_filter) {
                                if(lof_filter == 'END_TRUNC') return 'LOFTEE: End Truncation<br>This variant falls in the last 5% of the transcript.';
                                else if(lof_filter == 'INCOMPLETE_CDS') return 'LOFTEE: Incomplete CDS<br>The start or stop codons are not known for this transcript.';
                                //else if(lof_filter == 'NON_CAN_SPLICE_SURR') return 'LOFTEE: Non Canonical Splicing<br>This exon has surrounding splice sites that are non-canonical (not GT..AG).';
                                else if(lof_filter == 'EXON_INTRON_UNDEF') return 'LOFTEE: Exon-Intron Boundaries<br>The exon/intron boundaries of this transcript are undefined in the EnsEMBL API.';
                                else if(lof_filter == 'SMALL_INTRON') return 'LOFTEE: Small Intron<br>The LoF falls in a transcript whose exon/intron boundaries are undefined in the EnsEMBL API.';
                                else if(lof_filter == 'NON_CAN_SPLICE') return 'LOFTEE: Non Canonical Splicing<br>This variant falls in a non-canonical splice site (not GT..AG).';
                                else if(lof_filter == 'ANC_ALLELE') return 'LOFTEE: Ancestral Allele<br>The alternate allele reverts the sequence back to the ancestral state.';
                                 //else return "LOFTEE: " + lof_filter;
                            });
                        loftee_tooltip += lof_filters.join("<br>");
                    }

                    if(worst_vep_annotation.lof_flags == 'NAGNAG_SITE') {
                        loftee_tooltip += "LOFTEE: NAGNAG site<br>This acceptor site is rescued by another adjacent in-frame acceptor site.";
                    }

                    if (loftee_tooltip.length > 0) { %>
                        <span class="loftee label label-danger gotooltip" style="margin-left:10px; vertical-align:top; font-size:10px; color:white; display:inline" data-placement="top" data-original-title="<%= loftee_tooltip %>">
                            LC LoF
                        </span><br />
                    <%
                    }
                } %>

                <br/>

                <% if (worst_vep_annotation.hgvsc) { %>
                   <span>HGVS.C</span> <%= unescape(worst_vep_annotation.hgvsc.split(':').pop()) %></span><br />
                <% } %>
                <% if (worst_vep_annotation.hgvsp) { %>
                   <span>HGVS.P</span> <%= unescape(worst_vep_annotation.hgvsp.split(':').pop()) %></span><br />
                <% } %>

                <% var variantSearchLinks = utils.getVariantSearchLinks(variant); %>
                    <sup><a target="_blank" href="<%= variantSearchLinks['google'] %>">google</a> | <a target="_blank" href="<%= variantSearchLinks['pubmed'] %>">pubmed</a></sup>
                <% } %>
                <% if (variant.extras && variant.extras.svtype) { %>
                    <br /><br />
                    <span>VARIANT TYPE</span><br />
                     <%= variant.extras.svtype.replace("INS", "Insertion").replace("DEL", "Deletion") %> (<%= Math.abs(variant.extras.svlen) %>bp)<br/>
                <% } %>


        </div>

        <div class="cell predictions">
            <% if (variant.annotation) { %>
                <% if (variant.annotation.polyphen) { %>
                    <% if (variant.annotation.polyphen == 'probably_damaging') { %><i class="fa fa-circle redcircle"></i><% } %>
                    <% if (variant.annotation.polyphen == 'possibly_damaging') { %><i class="fa fa-circle yellowcircle"></i><% } %>
                    <% if (variant.annotation.polyphen == 'benign') { %><i class="fa fa-circle greencircle"></i><% } %>
                    <span>Polyphen</span> <%= variant.annotation.polyphen.replace("_", " ") %> <br/>
                <% } %>
                <% if (variant.annotation.sift) { %>
                    <% if (variant.annotation.sift == 'damaging') { %><i class="fa fa-circle redcircle"></i><% } %>
                    <% if (variant.annotation.sift == 'tolerated') { %><i class="fa fa-circle greencircle"></i><% } %>
                    <span>SIFT</span> <%= variant.annotation.sift %> <br/>
                <% } %>
                <% if (variant.annotation.muttaster) { %>
                    <% if (variant.annotation.muttaster == 'disease_causing') { %><i class="fa fa-circle redcircle"></i><% } %>
                    <% if (variant.annotation.muttaster == 'polymorphism') { %><i class="fa fa-circle greencircle"></i><% } %>
                    <span>Mut Taster</span> <%= variant.annotation.muttaster.replace("_", " ") %> <br/>
                <% } %>
                <% if (variant.annotation.fathmm) { %>
                    <% if (variant.annotation.fathmm == 'damaging') { %><i class="fa fa-circle redcircle"></i><% } %>
                    <% if (variant.annotation.fathmm == 'tolerated') { %><i class="fa fa-circle greencircle"></i><% } %>
                    <span>FATHMM</span> <%= variant.annotation.fathmm %> <br/>
                <% } %>
                <% if (variant.annotation.cadd_phred) {
                    var phred = parseFloat(variant.annotation.cadd_phred)
                    if( phred >= 20 ) { %>
                        <i class="fa fa-circle redcircle"></i>
                    <% } else if( phred >= 10 ) { %>
                        <i class="fa fa-circle yellowcircle"></i>
                    <% } else { %>
                        <i class="fa fa-circle greencircle"></i>
                    <% } %>
                    <span>CADD PHRED</span> <%= variant.annotation.cadd_phred %> <br/>
                <% } %>
                <% if (variant.annotation.dann_score) {
                    var dann_score = parseFloat(variant.annotation.dann_score)
                    if( dann_score >= 0.96 ) { %>
                        <i class="fa fa-circle redcircle"></i>
                    <% } else if( dann_score >= 0.93 ) { %>
                        <i class="fa fa-circle yellowcircle"></i>
                    <% } else { %>
                        <i class="fa fa-circle greencircle"></i>
                    <% } %>
                    <span>DANN SCORE</span> <%= parseFloat(variant.annotation.dann_score).toPrecision(2) %> <br/>
                <% } %>
                <% if (variant.annotation.revel_score) {
                    var revel_score = parseFloat(variant.annotation.revel_score)
                    if( revel_score >= 0.75 ) { %>
                        <i class="fa fa-circle redcircle"></i>
                    <% } else if( revel_score >= 0.5 ) { %>
                        <i class="fa fa-circle yellowcircle"></i>
                    <% } else { %>
                        <i class="fa fa-circle greencircle"></i>
                    <% } %>
                    <span>REVEL SCORE</span> <%= parseFloat(variant.annotation.revel_score).toPrecision(2) %> <br/>
                <% } %>
                <% if (variant.annotation.eigen_phred) {
                    var eigen_phred = parseFloat(variant.annotation.eigen_phred)
                    if( eigen_phred >= 1 ) { %>
                        <i class="fa fa-circle redcircle"></i>
                    <% } else if( eigen_phred >= 2 ) { %>
                        <i class="fa fa-circle yellowcircle"></i>
                    <% } else { %>
                        <i class="fa fa-circle greencircle"></i>
                    <% } %>
                    <span>EIGEN PHRED</span> <%= parseFloat(variant.annotation.eigen_phred).toPrecision(2) %> <br/>
                <% } %>
                <% if (variant.annotation.mpc_score) {
                    var mpc_score = parseFloat(variant.annotation.mpc_score)
                    if( mpc_score >= 2 ) { %>
                        <i class="fa fa-circle redcircle"></i>
                    <% } else if( mpc_score >= 1 ) { %>
                        <i class="fa fa-circle yellowcircle"></i>
                    <% } else { %>
                        <i class="fa fa-circle greencircle"></i>
                    <% } %>
                    <span>MPC SCORE</span> <%= parseFloat(variant.annotation.mpc_score).toPrecision(2) %> <br/>
                <% } %>
                <% if (variant.annotation.primate_ai_score) {
                    var primate_ai_score = parseFloat(variant.annotation.primate_ai_score)
                    if( primate_ai_score >= 0.7 ) { %>
                        <i class="fa fa-circle redcircle"></i>
                    <% } else if( primate_ai_score >= 0.5 ) { %>
                        <i class="fa fa-circle yellowcircle"></i>
                    <% } else { %>
                        <i class="fa fa-circle greencircle"></i>
                    <% } %>
                    <span>PRIMATE AI</span> <%= parseFloat(variant.annotation.primate_ai_score).toPrecision(2) %> <br/>
                <% } %>
                <% if (variant.annotation.splice_ai_delta_score) {
                    var splice_ai_delta_score = parseFloat(variant.annotation.splice_ai_delta_score)
                    if( splice_ai_delta_score >= 0.8 ) { %>
                        <i class="fa fa-circle redcircle"></i>
                    <% } else if( splice_ai_delta_score >= 0.5 ) { %>
                        <i class="fa fa-circle yellowcircle"></i>
                    <% } else { %>
                        <i class="fa fa-circle greencircle"></i>
                    <% } %>
                    <span>SPLICE AI</span> <%= parseFloat(variant.annotation.splice_ai_delta_score).toPrecision(2) %> <br/>
                <% } %>
            <% } %>
        </div>
        <div class="cell frequencies">
            <% if(variant.annotation && variant.annotation.freqs) {
                var grch37Coords = variant.extras.grch37_coords || (variant.chr+"-"+variant.pos+"-"+variant.ref+"-"+variant.alt);
                var grch37VariantIdSplit = grch37Coords.split("-");
                var grch37Pos = parseInt(grch37VariantIdSplit[1]);
                var grch37Region = grch37VariantIdSplit[0] + "-" + (grch37Pos - 100) + "-" + (grch37Pos + 100);

                var freqs = variant.annotation.freqs;
                if (variant.annotation.db == "elasticsearch") {
                    var g1kAF = freqs["1kg_wgs_popmax_AF"] || freqs["1kg_wgs_AF"];
                    var exacAF = freqs["exac_v3_popmax_AF"] || freqs["exac_v3_AF"];
                    var gnomadExomesAF = freqs["gnomad_exomes_popmax_AF"] || freqs["gnomad_exomes_AF"];
                    var gnomadGenomesAF = freqs["gnomad_genomes_popmax_AF"] || freqs["gnomad_genomes_AF"];
                    var topmedAF = freqs["topmed_AF"];

                    var callsetAF = freqs["AF"];

                    var popCounts = variant.annotation.pop_counts;

                    var grch38Coords = variant.extras.grch38_coords;
                    if (topmedAF == 0 && grch38Coords) {
                        var grch38VariantIdSplit = grch38Coords.split("-");
                        var grch38Pos = parseInt(grch38VariantIdSplit[1]);
                        var grch38Region = grch38VariantIdSplit[0] + "-" + (grch38Pos - 100) + "-" + (grch38Pos + 100);
                    }
                    var freqTooltipText = '<table>' +
                        '<tr>'+
                            '<td>this callset: </td>'+
                            '<td> AC=' + popCounts["AC"] + ', </td>'+
                            '<td> AN=' + (popCounts["AN"] != null ? popCounts["AN"] : "?") + '</td>'+
                        '</tr>'+
                        '<tr>'+
                            '<td>1kg WGS: </td>'+
                            '<td> AC=' + popCounts["1kg_AC"] + ', </td>'+
                            '<td> AN=' + (popCounts["1kg_AN"] != null ? popCounts["1kg_AN"] : "?") + '</td>'+
                        '</tr>'+
                        '<tr>'+
                            '<td>gnomad exomes: </td>'+
                            '<td> AC=' + popCounts["gnomad_exomes_AC"] + ', </td>'+
                            '<td> AN=' + (popCounts["gnomad_exomes_AN"] != null ? popCounts["gnomad_exomes_AN"] : "?") + '</td>'+
                         '</tr>' +
                        '<tr>'+
                            '<td>gnomad genomes: </td>'+
                            '<td> AC=' + popCounts["gnomad_genomes_AC"] + ', </td>'+
                            '<td> AN=' + (popCounts["gnomad_genomes_AN"] != null ? popCounts["gnomad_genomes_AN"] : "?") + '</td>'+
                        '</tr>' +
                        (variant.extras.genome_version == "38" ? ('<tr>'+
                            '<td>topmed: </td>'+
                            '<td> AC=' + popCounts["topmed_AC"] + ', </td>'+
                            '<td> AN=' + (popCounts["topmed_AN"] != null ? popCounts["topmed_AN"] : "?") + '</td>'+
                        '</tr></table>') : '');

                } else {
                    var g1kAF = freqs["1kg_wgs_phase3_popmax"] || freqs["1kg_wgs_phase3"];
                    var exacAF = freqs["exac_v3_popmax"] || freqs["exac_v3"];
                    var gnomadExomesAF = freqs["gnomad-exomes2_popmax"];
                    if ( gnomadExomesAF == null ) {
                        gnomadExomesAF = freqs["gnomad-exomes2"];
                    }
                    var gnomadGenomesAF = freqs["gnomad-genomes2_popmax"];
                    if ( gnomadGenomesAF == null ) {
                        gnomadGenomesAF = freqs["gnomad-genomes2"];
                    }
                    var popCounts = {};
                    var freqTooltipText = "";
                }
                %>
                <div class="gotooltip" data-placement="top" title="<%= freqTooltipText %>">
                    <% if (callsetAF != null) { %>
                        <div>
                          <span>THIS CALLSET</span>
                          <%= callsetAF.toPrecision(2) %>
                          <% if (popCounts["AC"] != null) { %>
                            <small style="margin-left: 10px;">
                              AC=<%= popCounts["AC"] %>
                              <% if (popCounts["AN"]) { %>
                                out of
                                <%= popCounts["AN"] %>
                              <% } %>
                            </small>
                          <% } %>
                          <% if (popCounts["Hom"] != null) { %>
                              <small style="margin-left: 10px">Hom=<%= popCounts["Hom"] %></small>
                          <% } %>
                        </div>
                    <% } %>

                    <% if (g1kAF != null) { %>
                        <div>
                            <span>1KG WGS</span><%= g1kAF.toPrecision(2) %>
                        </div>
                        <!-- 1kg Hom counts not available in dataset -->
                    <% } %>

                    <% if (exacAF != null) { %>
                        <div>
                            <span>EXAC</span>
                            <% if (exacAF > 0) { %>
                                <a target="_blank" href="http://exac.broadinstitute.org/variant/<%= grch37Coords %>"><%= exacAF.toPrecision(2) %> </a>
                            <% } else { %>
                                 <a target="_blank" href="http://exac.broadinstitute.org/region/<%= grch37Region %>">0.0 </a>
                            <% } %>
                            <% if (popCounts["exac_v3_Hom"] != null) { %>
                                <small style="margin-left: 10px">Hom=<%= popCounts["exac_v3_Hom"] %></small>
                                <% if (variant.chr.endsWith('X') && popCounts["exac_v3_Hemi"] != null) { %>
                                    <small style="margin-left:10px">Hemi=<%= popCounts["exac_v3_Hemi"] %></small>
                                <% } %>
                            <% } %>
                        </div>
                    <% } %>

                    <% if (gnomadExomesAF != null) { %>
                        <div>
                            <span>GNOMAD EXOMES</span>
                            <% if (gnomadExomesAF > 0) { %>
                                <a target="_blank" href="http://gnomad.broadinstitute.org/variant/<%= grch37Coords %>"><%= gnomadExomesAF.toPrecision(2) %> </a>
                            <% } else { %>
                                <a target="_blank" href="http://gnomad.broadinstitute.org/region/<%= grch37Region %>">0.0 </a>
                            <% } %>
                            <% if (popCounts["gnomad_exomes_Hom"] != null) { %>
                                <small style="margin-left: 10px">Hom=<%= popCounts["gnomad_exomes_Hom"] %></small>
                                <% if (variant.chr.endsWith('X') && popCounts["gnomad_exomes_Hemi"] != null) { %>
                                    <small style="margin-left:10px">Hemi=<%= popCounts["gnomad_exomes_Hemi"] %></small>
                                <% } %>
                            <% } %>
                        </div>
                    <% } %>

                    <% if (gnomadGenomesAF != null) { %>
                        <div>
                            <span>GNOMAD GENOMES</span>
                            <% if (gnomadGenomesAF > 0) { %>
                                <a target="_blank" href="http://gnomad.broadinstitute.org/variant/<%= grch37Coords %>"><%= gnomadGenomesAF.toPrecision(3) %></a>
                            <% } else { %>
                                <a target="_blank" href="http://gnomad.broadinstitute.org/region/<%= grch37Region %>">0.0</a>
                            <% } %>
                            <% if (popCounts["gnomad_genomes_Hom"] != null) { %>
                                <small style="margin-left: 10px">Hom=<%= popCounts["gnomad_genomes_Hom"] %></small>
                                <% if (variant.chr.endsWith('X') && popCounts["gnomad_genomes_Hemi"] != null) { %>
                                    <small style="margin-left:10px">Hemi=<%= popCounts["gnomad_genomes_Hemi"] %></small>
                                <% } %>
                            <% } %>
                        </div>
                    <% } %>
                    <% if (topmedAF != null && variant.extras.genome_version == "38") { %>
                        <div>
                            <span>TOPMED</span>
                            <% if (topmedAF > 0) { %>
                                <a target="_blank" href="https://bravo.sph.umich.edu/freeze5/hg38/variant/<%= grch38Coords %>"><%= topmedAF.toPrecision(3) %></a>
                            <% } else { %>
                                <a target="_blank" href="https://bravo.sph.umich.edu/freeze5/hg38/region/<%= grch38Region %>">0.0</a>
                            <% } %>
                            <% if (popCounts["topmed_Hom"] != null) { %>
                                <small style="margin-left: 10px">Hom=<%= popCounts["topmed_Hom"] %></small>
                            <% } %>
                        </div>
                    <% } %>
                </div>
            <% } %>
        </div>
        <% if (show_genotypes) { %>
            <div class="genotypes">
                <% _.each(individuals, function(individual) {
                    var indiv_id = individual.indiv_id;
                    var genotype=variant.genotypes[indiv_id];
                %>
                    <div class="cell geno-cell">
                        <div class="gotooltip" data-placement="top" title="<%= utils.getGenoMouseover(variant, indiv_id) %>">
                            <div class="genotype-icon-container">
                                <i class="fa <%= utils.getPedigreeIcon(individual_map[indiv_id]) %>"></i>
                                <% if (genotype_family_id) { %>
                                    <a href="/project/<%= individual.project_id %>/family/<%= individual.family_id %>/mendelian-variant-search" target="_blank">
                                        <% print(individual.family_id + ' / ') %>
                                    </a>
                                <% } %>
                                <%= individual.nickname || individual.indiv_id %>
                            </div>
                            <% if(genotype && genotype.alleles && genotype.alleles.length > 0 && genotype.num_alt != -1) {  %>
                                <% _.each(genotype.alleles, function(allele, i) { %>
                                    <% if (i>0) { print(' / ') } %>
                                    <span class="<% if (allele == variant.alt) { print('alt-allele ') } else if (allele != variant.ref) {print('other-allele')} %>">
                                        <%= allele.slice(0,3) %>
                                        <% if (allele.length > 3) { print('..') } %>
                                    </span>
                                <% }); %>
                            <% } else { %>
                                <span class="other-allele">no call</span>
                            <% } %><br>
                            <span class="gq">
                                <% if(genotype) { %>
                                    <%= genotype.gq %>
                                    <% if (genotype.filter != "pass") { %>
                                        <br />filter: <%= genotype.filter %>
                                    <% } %>
                                <% } %>
                            </span>
                        </div>
                        <% if(genotype && genotype.extras && genotype.extras.cnvs)  {
                            var cnvs = genotype.extras.cnvs;
                        %>
                            <span class="label label-danger gotooltip"
                                    data-placement="top"
                                    title="Copy Number: <%= cnvs['cn'] %><br>LRR median: <%= cnvs['LRR_median'] %><br>LRR stdev: <%= cnvs['LRR_sd'] %><br>SNPs supporting call: <%= cnvs['snps'] %><br>Size: <%= cnvs['size'] %><br>Found in: <% print(parseInt(cnvs['freq'])-1) %> other samples<br>Type: <%= cnvs['type'] %><br>Array: <%= cnvs['array'].replace(/_/g, ' ') %><br>Caller: <%= cnvs['caller'] %><br>">
                                CNV: <%= cnvs['cn'] > 2 ? 'Duplication' : 'Deletion' %>
                            </span><br>
                        <% } %>
                    </div>
                <% }); %>
            </div>
        <% } %>
        <% if (actions.length > 0) { %>
            <div class="cell actions" style="text-align:right">
                <% _.each(actions, function(action) { %>
                    <a class="btn btn-primary btn-xs action" data-action="<%= action.action %>"> <%= action.name %></a><br/>
                <% }); %>
            </div>
        <% } %>
    </div>
</script>
