<script type="text/template" id="tpl-cohort-gene-search-form">
    <h3 class="section-header">Inheritance</h3>
    <div id="tplholder-select-inheritance"></div>
    <h3 class="section-header">Variant Annotations</h3>
    <div id="tplholder-select-variants"></div>
    <h3 class="section-header">Call Quality</h3>
    <div id="select-quality-filter-container"></div>
</script>


<script type="text/template" id="tpl-cohort-variant-search-form">
    <h4 class="section-header">
        Genotypes To Search
    </h4>
    If left unselected, no inheritance filter will be applied. 
    <div id="tplholder-select-inheritance"></div>

    <h3 class="section-header">Variants <small>Select which variants you want to include</small></h3>

    <div id="tplholder-select-variants"></div>
    <div id="select-quality-filter-container"></div>
</script>


<script type="text/template" id="tpl-cohort-variant-search-results">
    <div id="summary-container">
        <div id='variant-stats-container'>
            <% if (num_variants == 0) { %>
                <div class="alert alert-danger">No variants found</div>
            <% } else { %>
                <div class="alert alert-success">
                    <strong><%= num_variants %></strong> variants found<br/>
                </div>
            <% } %>
        </div>
    </div>
    <div class="basic-variants-list leftalign"></div>
</script>


<script type="text/template" id="tpl-select-cohort-search-method">

    <div class="row">
        <div class="col-md-10 cohort-inheritance-container">
            <div class="inheritance-section">

                <label class="radio inheritance-mode">
                    <input type="radio" name="cohort_inheritance" value="recessive">
                    <strong>Recessive</strong>

                    <p>Genes with homozygous recessive, x-linked recessive, or compound heterozygous variants.</p>
                </label>

                <label class="radio inheritance-mode">
                    <input type="radio" name="cohort_inheritance" value="homozygous_recessive">
                    <strong>Homozygous Recessive</strong>

                    <p>Genes that include a homozygous alternate variant.</p>
                </label>

                <label class="radio inheritance-mode">
                    <input type="radio" name="cohort_inheritance" value="x_linked_recessive">
                    <strong>X-Linked Recessive</strong>

                    <p>Genes that include homozygous recessive variant on the X Chromosome. </p>
                </label>

                <label class="radio inheritance-mode">
                    <input type="radio" name="cohort_inheritance" value="compound_het">
                    <strong>Compound Heterozygous</strong>

                    <p>
                        Genes that contain a pair of heterozygous variants.
                        Currently we do not attempt to determine whether the two variants occur on the same haploptype,
                        as this is not possible in most cases with short reads exome data.
                    </p>
                </label>

                <label class="radio inheritance-mode">
                    <input type="radio" name="cohort_inheritance" value="dominant">
                    <strong>Dominant</strong>

                    <p>Genes with a single heterozygous variant</p>
                </label>

            </div>

        </div>

    </div>

</script>

<script type="text/template" id="tpl-cohort-results">
    <div class="col-md-12">
        <div id="summary-container">
            <div class="results-stats-container">
                <div class="alert alert-info">Returned <strong><%= genes.length %></strong> genes</div>
            </div>
        </div>
        <div id="resultsTableContainer">
            <table class="table table-fixed-header table-big" id="resultsTable">
                <thead class="header">
                <tr id="resultsTableHeaderRow">
                    <th>Gene</th>
                    <th>
                        # Hits
                        <i class="icon-question-sign icon-popover" title="Number of Hits" data-content="
                                Number of individuals in this cohort that display this inheritance with these variants.
                            "></i>
                    </th>
                    <th>
                        # Population Hits
                        <i class="icon-question-sign icon-popover" title="Population Hits" data-content="
                                This shows the number of individuals in a European control dataset (n=375)
                                that have this type of inheritance in this gene, with the same filters applied as above.
                            "></i>
                    </th>
                    <th>
                        p (Fisher)
                        <i class="icon-question-sign icon-popover" title="p value (Fisher's exact)" data-content="
                            p value from Fisher's exact test (two sided) when the number of hits in this cohort is compared to the
                            number of population hits.
                            "></i>
                    </th>
                    <th># Unique Variants</th>
                    <th>Position</th>
                    <th>
                        Coding Size (bases)
                        <i class="icon-question-sign icon-popover" title="Coding Size" data-content="
                                Sum of the length of all exons that contribute to a protein coding transcript.
                                Note that there is some discrepancy between this size and the size of the target capture region in this gene.
                            "></i>
                    </th>
                </tr>
                </thead>
                <tbody id="resultsTableBody"></tbody>
            </table>
        </div>
    </div>
</script>

<script type="text/template" id="tpl-cohort-single-gene-view">
    <td>
        <a class="gene-info-link" data-gene_id="<%= gene.gene_id %>">
            <%= gene.gene_name %>
        </a>
    </td>
    <td><%= gene.num_hits %></td>
    <td>
        <%= gene.control_comparison.control_hits %>
    </td>
    <td>
        <%= gene.control_comparison.fisher_2sided_pvalue.toPrecision(3) %>
    </td>
    <td>
        <a class="variants-link" data-gene_id="<%= gene.gene_id %>">
            <%= gene.num_unique_variants %>
        </a>
    </td>
    <td>
        <%if (gene.chr) { %>
            <%= gene.chr %>:<%= gene.start %>-<%= gene.end %>
        <% } %>
    </td>
    <td><% if (gene.stats) { print(gene.stats.coding_size); } %></td>
</script>

<script type="text/template" id="tpl-cohort-variants">

    <p>
        The table below shows all the variants in this gene that were considered as part of your search
        (ie. they passed the variant filter) where at least one of the individuals in this cohort contains at least one
        alternate allele.
        I want to note two implications of that statement:
    <ul>
        <li>There may be (and in fact likely are) additional variants in this gene that did not pass the variant filter
            but still may be important.
        </li>
        <li>
            Some of these variants might not contribute to the inheritance mode you specified.
        </li>
    </ul>
    </p>
    <div class="variants-container"></div>
    </div>

</script>

<script type="text/template" id="tpl-gene-variants">
    <div class="row-fluid">
        <div class="col-md-12">
            <div class="variants-table-container">
            </div>
        </div>
    </div>
</script>


<script type="text/template" id="tpl-cohort-quality-filter">
    <div class="row">
        <div class="col-md-3 left-select-container">
            <div class="filter-prefill-box">
                <p><em>Call Quality</em>: Initialize from one of our default variant quality criteria</p>
                <div class="row">
                    <div class="col-md-8 col-md-offset-2">
                        <select class="form-control col-md-2" id="quality-defaults-select">
                            <option value="">---</option>
                            <% _.each(defaultQualityFilters, function(qualityFilter) { %>
                            <option value="<%= qualityFilter.slug %>"><%= qualityFilter.name %></option>
                            <% }); %>
                        </select>
                    </div>
                </div>
            </div>
        </div>
        <div class="col-md-9 variant-select-controls">
            <div class="panel-group">
                <div class="panel panel-default">
                    <div class="panel-heading">
                        <a data-toggle="collapse" href="#collapse-qual">
                            Quality Metrics
                        </a>
                    </div>
                    <div id="collapse-qual" class="panel-body in panel-collapse">
                        <div class="row">
                            <div class="col-sm-4">
                                <div class="annot-section-title">FILTER value</div>
                                <div class="annot-section-content">
                                    <select class="form-control" name="filter-select" id="filter-select">
                                        <option value="">---</option>
                                        <option value="pass">PASS variants only</option>
                                    </select>
                                </div>
                            </div>
                            <div class="col-sm-4">
                                <div class="annot-section-title">Genotype Quality</div>
                                <div class="annot-section-content">
                                    <div id="gq-quality-container"></div>
                                </div>
                            </div>
                            <div class="col-sm-4">
                                <div class="annot-section-title">Allele Balance</div>
                                <div class="annot-section-content">
                                    <div id="ab-quality-container"></div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
                <div class="panel panel-default">
                    <div class="panel-heading">
                        <a data-toggle="collapse" href="#collapse-qual-sample">
                            In This Sample
                        </a>
                    </div>
                    <div id="collapse-qual-sample" class="panel-body in panel-collapse">
                        <div class="row">
                            <div class="col-sm-4">
                                <div class="annot-section-title">Het Ratio</div>
                                <div class="annot-section-content">
                                    <div id="het-ratio-container"></div>
                                </div>
                            </div>
                            <div class="col-sm-4">
                                <div class="annot-section-title">Hom Alt Ratio</div>
                                <div class="annot-section-content">
                                    <div id="hom-alt-ratio-container"></div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</script>


<script type="text/template" id="tpl-add-cohort">
    <div class="fullwidth-controls-container">
        <div class="fullwidth-controls">
            <div class="col">
                <a class="" id="select-with-phenotype">Select individuals with phenotype</a><br>
            </div>
            <div class="col">
                <a class="" id="select-from-list">Select individuals from list</a><br>
            </div>
            <div class="col">
                <a class="btn btn-primary" id="add-cohort-submit">Create Cohort From Selected</a>
            </div>
        </div>
    </div>
    <div id="individuals-table-container"></div>
    <div id="modal-base" class="modal fade" tabindex="-1" role="dialog">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-body" id="modal-inner">
                    <h3>hi</h3>
                </div>
            </div>
        </div>
    </div>
</script>


<script type="text/template" id="tpl-select-individuals-with-phenotype">
<div class="modal-header">
    <button type="button" class="close" data-dismiss="modal">×</button>
    <h3 id="myModalLabel">Select individuals iwth phenotype</h3>
</div>
<div class="modal-body">
    <p>Choose a phenotype and a value below, and families with that phenotype will be selected. </p>
    <p><em>Note: </em>right now can only choose "yes" as a binary value - others are temporarily disabled. </p>
    <div class="select-phenotype-container"></div>
    <a id="select-with-phenotype-submit" class="btn btn-primary">Select These Samples</a>
</div>
</script>


<script type="text/template" id="tpl-select-from-indiv-id-list">

<div class="modal-header">
    <button type="button" class="close" data-dismiss="modal">×</button>
    <h3 id="myModalLabel">Select From List</h3>
</div>
<div class="modal-body">
    <p>
        Copy/paste a list of <em>Individual IDs</em> into the field below.
        Click "Continue" they'll be selected in the previous table.
    </p>
    <div class="form-horizontal">
        <div class="control-group">
            <label class="control-label" for="id_name" >List of IDs</label>
            <div class="controls">
                <textarea id="select-from-indiv-id-list-textarea" placeholder="Paste one Individual ID per line..."></textarea>
            </div>
        </div>
        <div class="control-group">
            <div class="controls">
                <a id="select-from-indiv-id-list-submit" class="btn btn-primary">Continue</a>
            </div>
        </div>
    </div>
</div>

</script>
