<script type="text/template" id="tpl-gene-modal-content">


    <h4 class="section-header">Basics</h4>

    <dl class="dl-horizontal">

        <dt>Symbol</dt>
        <dd><%= gene.symbol %></dd>

        <dt>Ensembl ID</dt>
        <dd><%= gene.gene_id %></dd>

        <dt>Description</dt>
        <dd><%= gene.function_desc %></dd>

        <dt>Coordinates</dt>
        <dd>chr<%= gene.chrom %>:<%= gene.start %>-<%= gene.stop %></dd>

        <dt>Gene Type</dt>
        <dd><%= gene.gene_type %></dd>
    </dl>

    <h4 class="section-header">Stats</h4>
    <dl class="dl-horizontal">
        <dt>Coding Size</dt>
        <dd><%= (gene.coding_size/1000).toPrecision(2) %> kb</dd>
        <dt>Missense Constraint</dt>
        <dd>
            <% if (gene.tags.missense_constraint) { %>
            z-score: <%= gene.tags.missense_constraint.toPrecision(4) %> (ranked <%= gene.tags.missense_constraint_rank[0] %> most constrained out of <%= gene.tags.missense_constraint_rank[1] %> genes under study). &nbsp; &nbsp;  <i style='color:gray'>NOTE: Missense contraint is a measure of the degree to which the number of missense variants found in this gene in ExAC v0.3 is higher or lower than expected according to the statistical model described in [<a href='http://www.nature.com/ng/journal/v46/n9/abs/ng.3050.html'>K. Samocha 2014</a>]. In general this metric is most useful for genes that act via a dominant mechanism, and where a large proportion of the protein is heavily functionally constrained. For more details see this <a href='ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.3/functional_gene_constraint/README_forweb_cleaned_exac_r03_2015_03_16_z_data.txt'>README</a>.</i><br><br>
            <% } else { %>
            No score available
            <% } %>
        </dd>
        <dt>LoF Constraint</dt>
        <dd>
            <% if (gene.tags.lof_constraint) { %>
            pLI-score: <%= gene.tags.lof_constraint.toPrecision(4) %> (ranked <%= gene.tags.lof_constraint_rank[0] %> most intolerant of LoF mutations out of <%= gene.tags.lof_constraint_rank[1] %> genes under study). &nbsp; &nbsp;  <i style='color:gray'>NOTE: This metric is based on the amount of expected variation observed in the ExAC data and is a measure of how likely the gene is to be intolerant of loss-of-function mutations.</i>
            <% } else { %>
            No score available
            <% } %>
        </dd>
    </dl>

    <h4 class="section-header">Disease Associations</h4>
       <% if (gene.phenotype_info && gene.phenotype_info.has_mendelian_phenotype) { %>
            <div class="links-list-header">OMIM</div>
            <% for (var i=0; i < gene.phenotype_info.mim_phenotypes.length; i++) { %>
                <% if (gene.phenotype_info.mim_phenotypes[i].mim_id) { %>
                    <a target="_blank" href="http://www.omim.org/entry/<%= gene.phenotype_info.mim_phenotypes[i].mim_id %>">
                        <%=gene.phenotype_info.mim_phenotypes[i].description %>
                    </a>
                <% } else { %>
                    <%=gene.phenotype_info.mim_phenotypes[i].description %>
                <% } %>
                <br>
            <% } %>
	    <br>
	    <%= gene.disease_desc %>
        <% } else { %>
        <em>No disease associations</em>
        <% } %>
        <% if (gene.phenotype_info.orphanet_phenotypes.length > 0) { %>
            <div class="links-list-header">ORPHANET</div>
            <% for (var i=0; i < gene.phenotype_info.orphanet_phenotypes.length; i++) { %>
                <a target="_blank"
                   href="http://www.orpha.net/consor/cgi-bin/Disease_Search.php?lng=EN&data_id=20460&Disease_Disease_Search_diseaseGroup=<%= gene.phenotype_info.orphanet_phenotypes[i].orphanet_id %>">
                       <%= gene.phenotype_info.orphanet_phenotypes[i].description %>
                </a> <br>
            <% } %>
        <% } %>

    <h4 class="section-header">Links</h4>

    <dl class="dl-horizontal">

        <% if (gene.phenotype_info.mim_id != null) { %>
        <dt><a target="_blank" href="http://www.omim.org/entry/<%= gene.phenotype_info.mim_id %>">OMIM</a></dt>
        <dd>Database of Mendelian phenotypes</dd>
        <% } %>
        <dt><a target="_blank" href="http://www.ncbi.nlm.nih.gov/pubmed/?term=<%= gene.symbol %>">PubMed</a></dt><dd>Search PubMed for <%= gene.symbol %></dd>
        <dt><a target="_blank" href="http://www.genecards.org/cgi-bin/carddisp.pl?gene=<%= gene.symbol %>">GeneCards</a></dt><dd>Reference of public data for this gene</dd>
        <dt><a target="_blank" href="http://www.proteinatlas.org/<%= gene.gene_id %>/tissue">Protein Atlas</a></dt><dd>Detailed protein and transcript expression</dd>
	<dt><a target="_blank" href="http://www.ncbi.nlm.nih.gov/gene/?term=<%= gene.symbol %>">NCBI Gene</a></dt><dd>NCBI's gene information resource</dd>
	<dt><a target="_blank" href="http://www.gtexportal.org/home/gene/<%= gene.gene_id %>">GTEx Portal</a></dt><dd>Tissue-specific gene expression</dd>
	<dt><a target="_blank" href="http://monarchinitiative.org/search/<%= gene.gene_id %>">Monarch</a></dt><dd>Cross-species gene and phenotype resource</dd>
    </dl>

    <h4 class="section-header">Tissue-Specific Expression</h4>

    <p>This plot shows tissue-specific expression from GTEx release V6. These are normalized expression values with units of reads-per-kilobase-per-million (RPKMs) plotted on a log<sub>10</sub> scale, so that lower expression is to the left. 

    <div id="expression_plot"></div>
    <p id="no-gene-expression" style="display:none;">Expression data not available for this gene.</p>

</script>
